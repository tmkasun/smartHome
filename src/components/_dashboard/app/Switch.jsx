import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Icon } from '@iconify/react';
import bulbOutlined from '@iconify/icons-ant-design/bulb-outlined';
import bulbFilled from '@iconify/icons-ant-design/bulb-filled';
import acUnitIcon from '@iconify/icons-ic/baseline-ac-unit';
// material
import Tooltip from '@material-ui/core/Tooltip';
import { alpha, styled } from '@material-ui/core/styles';

import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import { LoadingButton } from '@material-ui/lab';
import { Card, Typography } from '@material-ui/core';
// custom

import { BorderLinearProgress, FacebookCircularProgress } from './components/SignalStrength';

// ----------------------------------------------------------------------

// ----------------------------------------------------------------------

const MIN = -90; // Minimum expected value
const MAX = -30; // Maximium expected value
// Function to normalise the values (MIN / MAX could be integrated)
const normalise = (value) => ((value - MIN) * 100) / (MAX - MIN);

const RootStyle = styled(Card)(({ theme }) => ({
  boxShadow: 'none',
  textAlign: 'center',
  padding: theme.spacing(5, 0),
  color: theme.palette.primary.darker,
  backgroundColor: theme.palette.primary.lighter
}));

/**
 * sdsadsa
 * @param {Object} props Sample
 * @returns Wola
 */
export default function Switch(props) {
  const { switchId, switchName } = props;
  const [switchInfo, setSwitchInfo] = useState(null);
  const [isLoading, setIsLoading] = useState(null);
  const [error, setError] = useState(null);
  const {
    seq: { data = {} }
  } = switchInfo || { seq: {} };
  const isOn = data.switch === 'on';
  // const classes = useStyles(isOn);

  const handleOnClick = (state) => {
    setIsLoading(true);
    let isOn = false;
    const nextSwitchState = { ...switchInfo };
    if (state === 'off') {
      isOn = true;
      nextSwitchState.seq.data.switch = 'on';
    } else {
      nextSwitchState.seq.data.switch = 'off';
    }
    const myHeaders = new Headers();
    myHeaders.append('Content-Type', 'application/json');

    const raw = JSON.stringify({
      isOn
    });

    const requestOptions = {
      method: 'POST',
      headers: myHeaders,
      body: raw,
      redirect: 'follow'
    };

    fetch(`https://home.knnect.com/switches/${switchId}/state`, requestOptions).finally(() =>
      setIsLoading(false)
    );
    setSwitchInfo(nextSwitchState);
  };
  const fetchData = () => {
    setIsLoading(true);
    const requestOptions = {
      method: 'GET',
      redirect: 'follow'
    };
    // https://www.npmjs.com/package/msw
    // fetch(`https://home.knnect.com/switches/${switchId}`, requestOptions)
    //   .then((response) => {
    //     if (!response.ok) {
    //       throw new Error('Something went wrong');
    //     }
    //     return response.json();
    //   })
    //   .then(setSwitchInfo)
    //   .catch((error) => {
    //     console.log('error', error);
    //     setError(error);
    //   })
    //   .finally(() => setIsLoading(false));
  };
  useEffect(fetchData, [switchId]);
  let icon;
  if (switchName === 'ac') {
    icon = acUnitIcon;
  } else {
    icon = isOn ? bulbFilled : bulbOutlined;
  }

  return (
    <RootStyle>
      <Button
        onClick={() => {
          handleOnClick(data.switch);
        }}
        sx={{
          margin: 'auto',
          display: 'flex',
          borderRadius: '50%',
          alignItems: 'center',
          width: 10,
          height: 10,
          backgroundColor: isOn ? 'white' : 'inherit',
          boxShadow: isOn ? '0px 0px 40px 20px #fff' : 'none',
          justifyContent: 'center',
          marginBottom: 3,
          color: (theme) => theme.palette.primary.dark,
          backgroundImage: (theme) =>
            `linear-gradient(135deg, ${alpha(theme.palette.primary.dark, 0)} 0%, ${alpha(
              theme.palette.primary.dark,
              0.24
            )} 100%)`
        }}
      >
        {isLoading ? (
          <FacebookCircularProgress />
        ) : (
          <Icon
            style={{ color: isOn ? '#f2ffe0' : 'inherit' }}
            icon={icon}
            width={54}
            height={54}
          />
        )}
      </Button>
      <Typography
        sx={{
          textTransform: 'capitalize'
        }}
        variant="h3"
      >
        {switchName}
      </Typography>
      <Tooltip title={`Signal strength (${data.signalStrength})`}>
        <Box sx={{ display: 'flex', alignItems: 'center' }}>
          <Box sx={{ width: '100%', mr: 1 }}>
            <BorderLinearProgress variant="determinate" value={normalise(data.signalStrength)} />
          </Box>
          <Box sx={{ minWidth: 35 }}>
            <Typography variant="body2" color="text.secondary">{`${Math.round(
              normalise(data.signalStrength)
            )}%`}</Typography>
          </Box>
        </Box>
      </Tooltip>
      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        Startup: {data.startup}
      </Typography>

      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        Device ID: {data.deviceid}
      </Typography>

      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        Puls: {data.pulse}
      </Typography>

      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        switch: {data.switch}
      </Typography>
      <LoadingButton
        onClick={() => {
          handleOnClick(data.switch);
        }}
        fullWidth
        size="large"
        type="submit"
        variant="contained"
        loading={isLoading}
        disabled={error}
      >
        {data.switch === 'off' ? 'On' : 'Off'}
      </LoadingButton>
    </RootStyle>
  );
}

Switch.propTypes = {
  /**
   Sample doc 1
   */
  switchId: PropTypes.string.isRequired,
  /**
   Sample doc 2
   */
  switchName: PropTypes.string
};
