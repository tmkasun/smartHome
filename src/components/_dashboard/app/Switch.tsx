import { useEffect, useState } from 'react';
import { Icon } from '@iconify/react';
import bulbOutlined from '@iconify/icons-ant-design/bulb-outlined';
import bulbFilled from '@iconify/icons-ant-design/bulb-filled';
import acUnitIcon from '@iconify/icons-ic/baseline-ac-unit';
// material
import Tooltip from '@mui/material/Tooltip';
import { alpha, styled } from '@mui/material/styles';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import { LoadingButton } from '@mui/lab';
import { Card, Typography } from '@mui/material';
// custom

import { BorderLinearProgress, FacebookCircularProgress } from './components/SignalStrength';

// ----------------------------------------------------------------------

const MIN = -90; // Minimum expected value
const MAX = -30; // Maximium expected value
// Function to normalise the values (MIN / MAX could be integrated)
const normalise = (value: number) => ((value - MIN) * 100) / (MAX - MIN);

const RootStyle = styled(Card)(({ theme }) => ({
  boxShadow: 'none',
  textAlign: 'center',
  padding: theme.spacing(5, 0),
  color: theme.palette.primary.darker,
  backgroundColor: theme.palette.primary.lighter,
}));

export default function Switch(props: { switchId: string; switchName: string }) {
  const { switchId, switchName } = props;
  const [switchInfo, setSwitchInfo] = useState(null);
  const [isLoading, setIsLoading] = useState<undefined | boolean>();
  const [error, setError] = useState(null);
  const {
    seq: { data = {} },
  }: { seq: { data: any } } = switchInfo || { seq: { data: {} } };
  const isOn = data.switch === 'on';

  const handleOnClick = (state: any) => {
    setIsLoading(true);
    let newIsOn = false;
    const nextSwitchState = { ...(switchInfo as any) };
    if (state === 'off') {
      newIsOn = true;
      nextSwitchState.seq.data.switch = 'on';
    } else {
      nextSwitchState.seq.data.switch = 'off';
    }
    const myHeaders = new Headers();
    myHeaders.append('Content-Type', 'application/json');

    const raw = JSON.stringify({
      isOn: newIsOn,
    });

    const requestOptions = {
      method: 'POST',
      headers: myHeaders,
      body: raw,
      redirect: 'follow',
    };

    fetch(`https://home.knnect.com/apis/switches/${switchId}/state`, requestOptions as any).finally(
      () => setIsLoading(false),
    );
    setSwitchInfo(nextSwitchState);
  };
  const fetchData = () => {
    setIsLoading(true);
    const requestOptions = {
      method: 'GET',
      redirect: 'follow',
    };
    fetch(`https://home.knnect.com/apis/switches/${switchId}`, requestOptions as any)
      .then((response) => {
        if (!response.ok) {
          throw new Error('Something went wrong');
        }
        return response.json();
      })
      .then((r) => {
        setSwitchInfo(r);
      })
      .catch((er) => {
        console.log('error', er);
        setError(er);
      })
      .finally(() => setIsLoading(false));
  };
  useEffect(fetchData, [switchId]);
  let icon;
  if (switchName === 'ac') {
    icon = acUnitIcon;
  } else {
    icon = isOn ? bulbFilled : bulbOutlined;
  }
  console.log(data);
  return (
    <RootStyle>
      <Button
        onClick={() => {
          handleOnClick(data.switch);
        }}
        sx={{
          margin: 'auto',
          display: 'flex',
          borderRadius: '50%',
          alignItems: 'center',
          width: 10,
          height: 10,
          backgroundColor: isOn ? 'white' : 'inherit',
          boxShadow: isOn ? '0px 0px 40px 20px #fff' : 'none',
          justifyContent: 'center',
          marginBottom: 3,
          color: 'primary.dark',
          backgroundImage: (theme) =>
            `linear-gradient(135deg, ${alpha(theme.palette.primary.dark, 0)} 0%, ${alpha(
              theme.palette.primary.dark,
              0.24,
            )} 100%)`,
        }}
      >
        {isLoading ? (
          <FacebookCircularProgress />
        ) : (
          <Icon
            style={{ color: isOn ? '#f2ffe0' : 'inherit' }}
            icon={icon}
            width={54}
            height={54}
          />
        )}
      </Button>
      <Typography
        sx={{
          textTransform: 'capitalize',
        }}
        variant="h3"
      >
        {switchName}
      </Typography>
      <Tooltip title={`Signal strength (${data.signalStrength})`}>
        <Box sx={{ display: 'flex', alignItems: 'center' }}>
          <Box sx={{ width: '100%', mr: 1 }}>
            <BorderLinearProgress variant="determinate" value={normalise(data.signalStrength)} />
          </Box>
          <Box sx={{ minWidth: 35 }}>
            <Typography variant="body2" color="text.secondary">{`${Math.round(
              normalise(data.signalStrength),
            )}%`}</Typography>
          </Box>
        </Box>
      </Tooltip>
      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        Startup: {data.startup}
      </Typography>

      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        Device ID: {data.deviceid}
      </Typography>

      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        Puls: {data.pulse}
      </Typography>

      <Typography variant="subtitle2" sx={{ opacity: 0.72 }}>
        switch: {data.switch}
      </Typography>
      <LoadingButton
        onClick={() => {
          handleOnClick(data.switch);
        }}
        fullWidth
        size="large"
        type="submit"
        variant="contained"
        loading={isLoading}
        disabled={error as unknown as boolean}
      >
        {data.switch === 'off' ? 'On' : 'Off'}
      </LoadingButton>
    </RootStyle>
  );
}
