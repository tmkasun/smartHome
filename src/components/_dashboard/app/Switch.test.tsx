import React from 'react';
import { render, waitForElementToBeRemoved, screen } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { OpenAPIBackend } from 'openapi-backend'; // https://github.com/anttiviljami/openapi-backend https://dev.to/epilot/testing-react-with-jest-and-openapi-mocks-8gc

import Switch from './Switch';

const api = new OpenAPIBackend({
  definition: '/home/kasun/projects/smartHome/swagger.yaml',
});

api.register('notFound', (c, res, ctx) => res(ctx.status(404)));
api.register('notImplemented', async (c, res, ctx) => {
  const { status, mock } = await api.mockResponseForOperation(
    (c.operation && c.operation.operationId) || 'NO-OP',
  );
  return res(ctx.status(status), ctx.json(mock));
});
const server = setupServer(
  rest.get('https://home.knnect.com/apis/*', (req, res, ctx) => {
    // res(ctx.json({ greeting: 'hello there' }));
    req.path = req.url.pathname.replace('/apis', '');
    return api.handleRequest(req, res, ctx);
  }),
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterEach(() => server.close());

test('Sonoff switch default', async () => {
  render(<Switch switchId={12} switchName="testSW" />);
  screen.getByTestId('fb-progress');
  expect(
    screen.getByRole('button', {
      name: /off/i,
    }),
  ).toBeDisabled();
  await waitForElementToBeRemoved(() => screen.getByTestId('fb-progress'));
  expect(
    screen.getByRole('heading', {
      name: /device id: 10011c4ca4/i,
    }),
  ).toBeVisible();
  expect(
    screen.getByRole('button', {
      name: /on/i,
    }),
  ).not.toBeDisabled();
  // screen.getByRole('heading', {
  //   name: /device id:/i,
  // });
});
