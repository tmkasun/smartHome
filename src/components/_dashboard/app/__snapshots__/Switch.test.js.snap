// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Switch Default 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1hs79rk-MuiPaper-root-MuiCard-root"
>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-qzfj3p-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <div
      className="MuiBox-root css-79elbk"
    >
      <span
        aria-valuenow={100}
        className="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorPrimary css-nt50a2-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "transform": "rotate(-90deg)",
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20}
            strokeWidth={4}
            style={
              Object {
                "strokeDasharray": "125.664",
                "strokeDashoffset": "0.000px",
              }
            }
          />
        </svg>
      </span>
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-cvhkgz-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate MuiCircularProgress-circleDisableShrink css-79nvmn-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20}
            strokeWidth={4}
            style={Object {}}
          />
        </svg>
      </span>
    </div>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <h3
    className="MuiTypography-root MuiTypography-h3 css-1g3mzzc-MuiTypography-root"
  >
    kasunTest
  </h3>
  <div
    aria-label="Signal strength (undefined)"
    aria-labelledby={null}
    className="MuiBox-root css-70qvj9"
    data-mui-internal-clone-element={true}
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="MuiBox-root css-1bhivts"
    >
      <span
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={NaN}
        className="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-1l6rwmg-MuiLinearProgress-root"
        role="progressbar"
      >
        <span
          className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style={
            Object {
              "transform": "translateX(NaN%)",
            }
          }
        />
      </span>
    </div>
    <div
      className="MuiBox-root css-pazuet"
    >
      <p
        className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        NaN%
      </p>
    </div>
  </div>
  <h6
    className="MuiTypography-root MuiTypography-subtitle2 css-imynep-MuiTypography-root"
  >
    Startup: 
  </h6>
  <h6
    className="MuiTypography-root MuiTypography-subtitle2 css-imynep-MuiTypography-root"
  >
    Device ID: 
  </h6>
  <h6
    className="MuiTypography-root MuiTypography-subtitle2 css-imynep-MuiTypography-root"
  >
    Puls: 
  </h6>
  <h6
    className="MuiTypography-root MuiTypography-subtitle2 css-imynep-MuiTypography-root"
  >
    switch: 
  </h6>
  <button
    className="MuiLoadingButton-root MuiLoadingButton-loading MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButtonBase-root Mui-disabled css-sppx0e-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="submit"
  >
    <div
      className="MuiLoadingButton-loadingIndicator MuiLoadingButton-loadingIndicatorCenter css-1yt7yx7-MuiLoadingButton-loadingIndicator"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 16,
            "width": 16,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
    </div>
    Off
  </button>
</div>
`;
